js-extensions/src/ClassMetadataStore.js

```diff
@@ -0,0 +1,121 @@
+     * NOTE: This ALSO tests the prototype hierarchy if objectToTest
+     * is a Function.
+     */
+    static isType(objectToTest, type) {
+        var o = objectToTest;
+        if (typeof o === type) {
+            return true;
+        }
+        if (type === String || type === 'string') {
+            return o instanceof String;
+        }
+        if (type === Function || type === 'function') {
+            return o instanceof Function;
+        }
+        if (type === Object || type === 'object') {
+            return o instanceof Object;
+        }
+        if (objectToTest instanceof Function) {
+            var proto = objectToTest;
+            while (proto) {
+                if (proto === type) {
+                    return true;
+                }
+                proto = Object.getPrototypeOf(proto);
+            }
+        }
+        return objectToTest instanceof type;
+    }
+    
+    componentType(componentType) {
+        return (extensions, onload) => {
+            extensions = extensions.filter(e => ClassMetadataStore.isType(e.instance, componentType));
+            onload(extensions);
+        };
+    }
+    
+    dataType(dataType) {
+        return (extensions, onload) => {
+            if (dataType && typeof(dataType) === 'object'
+                    && '_class' in dataType) { // handle the common API incoming data
+                dataType = dataType._class;
+            }
+            
+            this.getClassMetadata(dataType, (currentTypeInfo) => {
+                // prevent returning extensions for the given type
+                // when a more specific extension is found
+                var matchingExtensions = [];
+                eachType: for (var typeIndex = 0; typeIndex < currentTypeInfo.classes.length; typeIndex++) {
+                    // currentTypeInfo.classes is ordered by java hierarchy, including
+                    // and beginning with the current data type
+                    var type = currentTypeInfo.classes[typeIndex];
+                    for (var i = 0; i < extensions.length; i++) {
+                        var extension = extensions[i];
+                        if (type === extension.dataType) {
+                            matchingExtensions.push(extension);
+                        }
+                    }
+                    // if we have this specific type handled, don't
+                    // proceed to parent types
+                    if (matchingExtensions.length > 0) {
+                        break eachType;
+                    }
+                }
+                onload(matchingExtensions);
+            });
+        };
+    }
+
+    /**
+     * Returns a filtering function to only return untyped extensions
+     */
+    untyped() {
+        return (extensions, onload) => {
+            // exclude typed extensions when types not requested
+            extensions = extensions.filter(m => !('dataType' in m));
+            onload(extensions);
+        };
+    }
```

Comment:
Do `componentType`, `dataType` and `untyped` have anything to do with `ClassMetadataStore` ? Seems like they belong more on `ExtensionStore`, seeing as though they are for filtering the extensions in their.

Is there scope here to formalise the `Filter` mechanism into a type and extend it to create `ComponentTypeFilter`, `DataTypeFilter` etc in class files of their own?